extensions/kafka-eight-simpleConsumer/src/main/java/io/druid/firehose/kafka/KafkaSimpleConsumer.java

```diff
@@ -0,0 +1,402 @@
+        consumer = new SimpleConsumer(broker.host, broker.port, soTimeout, bufferSize, clientId + "leaderLookup");
+        TopicMetadataResponse resp = consumer.send(new TopicMetadataRequest(Collections.singletonList(topic)));
+
+        List<TopicMetadata> metaData = resp.topicsMetadata();
+        for (TopicMetadata item : metaData) {
+          if (topic.equals(item.topic())) {
+            for (PartitionMetadata part : item.partitionsMetadata()) {
+              if (part.partitionId() == partitionId) {
+                return part;
+              }
+            }
+          }
+        }
+      }
+      catch (Exception e) {
+        ensureNotInterrupted(e);
+        log.warn(
+            e, "error communicating with Broker [%s] to find leader for [%s] - [%s]", broker, topic, partitionId
+        );
+      }
+      finally {
+        if (consumer != null) {
+          try {
+            consumer.close();
+          }
+          catch (Exception e) {
+          }
+        }
+      }
+    }
+
+    return null;
+  }
+
+  private Broker findNewLeader(Broker oldLeader) throws InterruptedException
+  {
+    long retryCnt = 0;
+    while (true) {
+      PartitionMetadata metadata = findLeader();
+      if (metadata != null) {
+        replicaBrokers.clear();
+        for (Broker replica : metadata.replicas()) {
+          replicaBrokers.add(new KafkaBroker(replica.host(), replica.port()));
+        }
+
+        log.debug("findNewLeader - metadata[%s], previous leader[%s]", metadata, oldLeader);
+        Broker newLeader = metadata.leader();
+        if (newLeader != null) {
+          // We check the retryCnt here as well to make sure that we have slept a little bit
+          // if we don't notice a change in leadership.  Note, that this means that this method
+          // is hopefully only being called when we believe that leadership has changed.  Otherwise
+          // there is a big chunk of sleep being added for no reason.
+          if (oldLeader == null || !isCurrLeader(newLeader) || retryCnt != 0) {
+            return newLeader;
+          }
+        }
+      }
+
+      Thread.sleep(1000L);
+      retryCnt++;
+      // if could not find the leader for current replicaBrokers, let's try to
+      // find one via allBrokers
+      if (retryCnt >= 3 && (retryCnt - 3) % 5 == 0) {
+        log.warn("cannot find leader for [%s] - [%s] after [%s] retries", topic, partitionId, retryCnt);
+        replicaBrokers.clear();
+        replicaBrokers.addAll(allBrokers);
+      }
+    }
+  }
+
+  private boolean isCurrLeader(Broker broker) {
+    return (leaderBroker == null && broker == null) ||
+           leaderBroker.host().equalsIgnoreCase(broker.host()) && leaderBroker.port() == broker.port();
+  }
+
+  private void ensureNotInterrupted(Exception e) throws InterruptedException
+  {
+    if (Thread.interrupted()) {
+      log.info(e, "Interrupted during fetching for %s - %s", topic, partitionId);
```

Comment:
This is a normal case, because then the client of KafkaSimpleConsumer wants to stop it, it will interrupt the thread.